clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -491,7 +491,13 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
 
                 /* if channel is ready write to any sockets that have space in their buffer and for which we have data */
                 if (channel.ready() && key.isWritable()) {
-                    Send send = channel.write();
+                    Send send = null;
+                    try {
+                        send = channel.write();
+                    } catch (Exception e) {
+                        this.failedSends.add(channel.id());
```

Comment:
@parafiend Looking at this again, adding to `failedSends` in `poll` can result in multiple disconnect notiifcations for a channel (we have to guarantee exactly one). `failedSends` are processed on the following `poll()`. But for the write exception here, we process `close()` in the catch block and the channel is added to `disconnected` list. In the next poll, when `failedSends` are processed, the channel will be added again to `disconnected` list. It would be better to set a flag rather than update `failedSends` and change the close in the catch block to:
```
close(channel, !sendFailed);
```
This will close the channel in the current poll without waiting to process any outstanding requests.